x/sync/sync_test.go

```diff
@@ -793,68 +759,257 @@ func TestFindNextKeyRandom(t *testing.T) {
+			},
+		},
+		{
+			name: "change proof bad response - removed key from middle of response",
+			changeProofClient: func(db merkledb.MerkleDB) *p2p.Client {
+				handler := newFlakyChangeProofHandler(t, db, func(response *merkledb.ChangeProof) {
+					i := rand.Intn(max(1, len(response.KeyChanges)-1)) // #nosec G404
+					_ = slices.Delete(response.KeyChanges, i, min(len(response.KeyChanges), i+1))
+				})
+
+				return p2ptest.NewClient(t, context.Background(), handler, ids.GenerateTestNodeID(), ids.GenerateTestNodeID())
+			},
+		},
+		{
+			name: "all proof keys removed from response",
+			changeProofClient: func(db merkledb.MerkleDB) *p2p.Client {
+				handler := newFlakyChangeProofHandler(t, db, func(response *merkledb.ChangeProof) {
+					response.StartProof = nil
+					response.EndProof = nil
+				})
+
+				return p2ptest.NewClient(t, context.Background(), handler, ids.GenerateTestNodeID(), ids.GenerateTestNodeID())
+			},
+		},
+		{
+			name: "flaky change proof client",
+			changeProofClient: func(db merkledb.MerkleDB) *p2p.Client {
+				return p2ptest.NewClient(t, context.Background(), &flakyHandler{
+					Handler: NewGetChangeProofHandler(logging.NoLog{}, db),
+					c:       &counter{m: 2},
+				}, ids.GenerateTestNodeID(), ids.GenerateTestNodeID())
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+
+			ctx := context.Background()
+			dbToSync, err := generateTrie(t, r, 3*maxKeyValuesLimit)
+			require.NoError(err)
+
+			syncRoot, err := dbToSync.GetMerkleRoot(ctx)
+			require.NoError(err)
+
+			db, err := merkledb.New(
+				ctx,
+				memdb.New(),
+				newDefaultDBConfig(),
+			)
+			require.NoError(err)
+
+			var (
+				rangeProofClient  *p2p.Client
+				changeProofClient *p2p.Client
+			)
+
+			rangeProofHandler := NewGetRangeProofHandler(logging.NoLog{}, dbToSync)
+			rangeProofClient = p2ptest.NewClient(t, ctx, rangeProofHandler, ids.GenerateTestNodeID(), ids.GenerateTestNodeID())
+			if tt.rangeProofClient != nil {
+				rangeProofClient = tt.rangeProofClient(dbToSync)
+			}
+
+			changeProofHandler := NewGetChangeProofHandler(logging.NoLog{}, dbToSync)
+			changeProofClient = p2ptest.NewClient(t, ctx, changeProofHandler, ids.GenerateTestNodeID(), ids.GenerateTestNodeID())
+			if tt.changeProofClient != nil {
+				changeProofClient = tt.changeProofClient(dbToSync)
+			}
+
+			syncer, err := NewManager(ManagerConfig{
+				DB:                    db,
+				RangeProofClient:      rangeProofClient,
+				ChangeProofClient:     changeProofClient,
+				TargetRoot:            syncRoot,
+				SimultaneousWorkLimit: 5,
+				Log:                   logging.NoLog{},
+				BranchFactor:          merkledb.BranchFactor16,
+			}, prometheus.NewRegistry())
+
+			require.NoError(err)
+			require.NotNil(syncer)
+
+			// Start syncing from the server
+			require.NoError(syncer.Start(ctx))
+
+			// Simulate writes on the server
+			// TODO more than a single write when API is less flaky
```

Comment:
Yea good callout - I actually had to re-inspect this test to figure out what the race condition was. Added a comment here, but basically `UpdateSyncTarget` errors if it has already finished syncing (i.e it the syncer closes once it hits the target it's trying to hit). I noticed on larger `N` of writes that we were erroring out sometimes because we would finish a previous sync target before we updated it.